[dynamo] Enforce some invariants on `ConstantVariable.create`

StrongerXi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140984
* #141504

This addresses https://github.com/pytorch/pytorch/pull/140745#issuecomment-2480854259.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi commented on torch/_dynamo/variables/constant.py:
```diff
@@ -37,18 +36,29 @@
                 item_source = GetItemSource(source, i) if source else None
-                if item_source:
-                    install_guard(item_source.make_guard(GuardBuilder.CONSTANT_MATCH))
                 items.append(
```
I can't even figure out a test for this path, because it never fires for tests in `test/dynamo`...

jansel replied:
If you think it is dead code maybe try removing it.

StrongerXi commented on torch/_dynamo/variables/builder.py:
```diff
@@ -682,6 +682,11 @@ def build_key_value(i, k, v):
         elif istype(value, frozenset) and (
             ConstantVariable.is_literal(x) for x in value
         ):
+            # TODO: the `is_literal` check above is futile, because we are
+            # missing a `all` call around this generator object; this means we
+            # always wrap a fronzent with 1 layer of `ConstantVariable`, which
+            # interacts in a subtle way with `ConstantVariable.create`.
+            #
```
I tried to remove this branch in favor of `ConstantVariable.create`, but failed because
1. The branch was added for `test_frozenset_torch_func_contains`, which contains a frozenset of function objects.
2. `ConstantVariable.is_literal` fails for function objects
3. Even if we fix (2), we don't want `ConstantVariable.create` to end up using `ConstantVariable` for the item function, because it might get loaded and called later, as in `test_frozenset_torch_func_contains`.

It's unclear whether the lack of `all(...)` call here is intended. I _feel_ like it's a bug that happens to work for `test_frozenset_torch_func_contains`.

But let's say we fix (3) as well by using `LazyVariableTracker` or `VariableTracker.build` for the nested items. Then why do we even need `ConstantVariable` at all, why not just replace all uses of it with `LazyVariableTracker` or `VariableTracker.build`...?

jansel replied:
This looks like a bug to me, we should add the `all()` and fix and bugs that introduces.

StrongerXi replied:
Addressed in #141504.

StrongerXi commented on torch/_dynamo/variables/constant.py:
```diff
@@ -37,18 +36,29 @@
+        """
+        Create a `ConstantVariable` based on the given value.
+
+        NOTE: the caller must install the proper guards if needed; most often
+        the guard will be `CONSTANT_MATCH`.
```
Existing use cases of this (that has a source) all have `CONSTANT_MATCH` installed explicitly, except for 1 case which I fixed in this patch.

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Enforce some invariants on `ConstantVariable.create`"


This addresses https://github.com/pytorch/pytorch/pull/140745#issuecomment-2480854259.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Enforce some invariants on `ConstantVariable.create`"


This addresses https://github.com/pytorch/pytorch/pull/140745#issuecomment-2480854259.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi commented:
@pytorchbot merge

StrongerXi commented:
@pytorchbot merge
